### PR TITLE
RGB(A) support

### PR DESCRIPTION
**Pull request by billbrod**
Adds better support for using `imshow` to show RGB(A) images.

### CODE REVIEW

**Comment by billbrod** [comment]
Something I'd like feedback on: currently the RGBA has to be on the last dimension (like in matplotlib). Is that okay, or do we want to support more flexibility? For tensors in `plenpotic`, RGBA will be along the channel dimension (I'm about to push that), but the extra flexibility here might be confusing.

So the input can be:
 - 2d array (single grayscale image)
 - 3d array (multiple grayscale images)
 - 3d array (single RGB(A) image, `image.shape[-1] in [3,4]`)
 - list of 2d arrays (multiple grayscale images)
 - list of 3d arrays (multiple RGB(A) images, `image.shape[-1] in [3,4]`)
 - 4d array (multiple RGB(A) images, `image.shape[-1] in [3,4]`)

I'm pretty sure there's only one potential case of ambiguity: if you pass a 3d array with shape `(i, j, 3)` or `(i, j, 4)`. We can't tel if it's a single `i*j` color image or `i` `j*3` or `j*4` grayscale images. We have a new argument `as_rgb` (default `True`), which forces us one way or the other on this. It will raise Exceptions when it's `False` and it can't figure out how to make something a grayscale image, but not when it's `True` and it can't figure out how to make something an RGB(A) image (it will just plot in grayscale; because I want the default to do what you want most of the time).

**Comment by nikparth** [comment]
One thought I have is maybe it would be a lot simple to remove the 3d array functionality for multiple grayscale images and remove the 4d array for multiple RGB(A) images? 
That might get rid of a lot of the confusion.. Just force the user to make it a list whenever you need to display multiple images. 
Moreover, it's easy to write a helper function to do this list conversion i.e. if x is your image, then: `list_x = [x[...,i] for i in range(x.shape[-1])`

**Comment by billbrod** [comment]
In practice, it isn't actually that difficult to support both lists and the corresponding arrays, and I do find it useful as a user. It does make the docs a little confusing, but I think the only thing it would clear up is the ambiguous case I outlined above -- which is relatively uncommon (we don't often plot something that's `j*3` or `j*4`, so the assumption that it's a color image is fine). Have you had times where it's been confusing / unclear?

**Comment by nikparth** [comment]
My only confusion is that reading the docs I had to read them a few times to figure out all the scenarios, so I'm just in general for simpler functions but if the flexibility is very useful then that's totally cool. Another question..  in the current scenario, what happens if you provide an `(i,j,3)` image where it is not RGB but just 3 channels (i.e. LUV)? If you set as_rgb=False, does it just display these channels grayscale?

**Comment by billbrod** [comment]
Yeah, the docs are confusing. There's got to be a better way to explain it -- would really appreciate help on that.

It would display them as `i` `j*3` grayscale images right now. What you'd probably want is 3 `i*j` grayscale images, which you'd pass a `(3,i,j)` array to get. `pt.imshow` (which I'm writing the tests for) will take `(b,c,h,w)` tensors and can display RGB images using channel or display them as separate images, which is probably the more common use case

**Comment by nikparth** [comment]
Yeah I can try to help with the docs but my concern is that it's just so much flexibility that you still need to explain each option which is going to be confusing regardless.. Also I'm not really sure what the `i` `j*3` images means? what's the `(h,w)` for that? That seems to be the most confusing example to me so is supporting this option necessary? That may clear things up..

**Comment by billbrod** [comment]
True, but it feels weird to not support a 3d array. That's the most obvious thing to pass to `imshow`.

Sorry, I meant you have `i` images with height `j` and width `3`. Honestly, it is a bit of an edge case, but it would be weird to not support showing images of a certain size.

**Comment by nikparth** [comment]
Wait now im confused- what is the ambiguity then? if you set as_rgb = True, then all images are displayed grayscale and if as_rgb = False and the last dim is 3 or 4 it displays as RGB.. I'm not sure what you mean by ": if you pass a 3d array with shape (i, j, 3) or (i, j, 4). We can't tel if it's a single i*j color image or i j*3 or j*4 grayscale images." This should just be determined by the as_rgb argument?

**Comment by billbrod** [comment]
Yes, it is! If we didn't have the `as_rgb` arg, then it would be ambiguous -- that's the reason I added it. 

Also, it will raise an exception if you set `as_rgb=False` and it can't figure out a way to display grayscale images (e.g., it's a 4d array)

**Comment by nikparth** [comment]
Oh lol so then actually I dont see any issue then.. but I'll take a crack at rephrasing maybe.

**Comment by billbrod** [comment]
Haha cool. Yeah clearly it needs some rephrasing given how confusing it was!

File: pyrtools/tools/display.py [hunk 1]
```diff
@@ -351,29 +359,40 @@ def check_shape_1d(shapes):
...
-        all images will be automatically rescaled so they're displayed at the same size. thus,
-        their sizes must be scalar multiples of each other.
+        should be a 2d array (one image to display), 3d array (multiple
+        grayscale images to display, images are indexed along the first
+        dimension, or a single RGB(A) image, with RGB(A) along the final
+        dimension), list of 2d or 3d arrays (grayscale or RGB(A) images,
+        respectively), or 4d array (multiple RGB(A) images). the image(s) to be
+        shown. all images will be automatically rescaled so they're displayed
+        at the same size. thus, their sizes must be scalar multiples of each
+        other.
```
Comment (nikparth) [inline pyrtools/tools/display.py:395]: rephrasing: `np.array` or `list` (of np.array). If grayscale, the as_rgb argument (see below) must be set to True else, must be set to False. For grayscale, images are 2d array of shape (h,w), or 3d array of shape (n,h,w) where images are indexed on first dimension. You can also a provide a list of 2d arrays. For RGB(A) images, the image can be a 3d array of shape (h,w,3), 4d array of shape (n, h,w,3) or list of 3d arrays.  --- something like that so that it's emphasized that the as_rgb argument must be set correctly and then clearly delineate sentences for  grayscale vs. rgb sizing.

Comment (billbrod) [reply pyrtools/tools/display.py:395]: I like that. 

Right now, if `as_rgb=True` and it's a grayscale image, it won't fail, it will properly plot them as grayscale (since that's the default value). Should I make it default to `False` and fail if `as_rgb=True` and it's grayscale? (So, it fails if `as_rgb=False` and it can't figure out how to plot it as grayscale, but not if `as_rgb=True` and it can't figure out how to plot it as RGB(A))

Comment (nikparth) [reply pyrtools/tools/display.py:395]: I would say it should fail both if `as_rgb=True` and it's a grayscale image, and if `as_rgb=False` and it can't plot it as RGB. Basically the image input and that argument should always match or it should fail.

File: pyrtools/tools/display.py [hunk 2]
```diff
@@ -351,95 +359,123 @@ def check_shape_1d(shapes):
...
 
+    """
     if plot_complex not in ['rectangular', 'polar', 'logpolar']:
-        warnings.warn("Don't know how to handle plot_complex value %s, defaulting to "
-                      "'rectangular'" % plot_complex)
-        plot_complex = 'rectangular'
+        raise Exception("Don't know how to handle plot_complex value "
+                        f"{plot_complex}!")
 
     try:
```
Comment (ravos) [inline pyrtools/tools/display.py:472]: The current code might report a random error if the argument `image` is neither a `list` nor a `np.ndarray`. I'd recommend this:
```python
if type(image) is np.ndarray:
# not needed since it's checked later.
#    if image.ndim not in [2, 3]:
#        raise ValueError("unexpected dimension")
#    else:
    image = [image]

if type(image) is not list:
    raise TypeError('must be a list')    
```

File: pyrtools/tools/display.py [hunk 3]
```diff
@@ -461,6 +497,16 @@ def imshow(image, vrange='indep1', zoom=1, title='', col_wrap=None, ax=None,
     image_tmp = []
     title_tmp = []
     for img, t in zip(image, title):
+        if as_rgb and (img.ndim == 2 or img.shape[-1] not in [3, 4]):
```
Comment (ravos) [inline pyrtools/tools/display.py:500]: I'm wondering if `as_rgb = (img.shape[-1] in [3, 4])` is 100% correct? If so, it is possible to reduce another parameter.

Comment (ravos) [reply pyrtools/tools/display.py:500]: sorry `as_rgb = (img.ndim == 3) and (img.shape[-1] in [3, 4])`, and assuming there is no user mistake.

File: pyrtools/tools/display.py (hunk 2 above)
Comment (billbrod) [reply pyrtools/tools/display.py]: Ah, that's a good idea. I wrote it the way I did to try and make it more "duck typing", in case the user passes an object that can be converted easily to an array. Like `torch.tensor`, for example: there are cases where passing a tensor directly to imshow works as you'd like, which the above rules out. But given that we're also implementing `plenoptic.imshow`, which will handle tensors, that's probably no longer necessary...

File: pyrtools/tools/display.py (hunk 3 above)
Comment (billbrod) [reply pyrtools/tools/display.py]: That's true, we can probably get rid of that. I think it would be `as_rgb = (img.shape[-1] in [3,4] and img.ndim==3)` (because if someone passes an image with shape `(4,4)`, that should be a grayscale image). 

Currently, we don't allow mixing of RGB and grayscale images, but if we do something like the above to automatically infer whether each image should be RGB or not, then we should be able to.

File: pyrtools/tools/display.py [hunk 4]
```diff
@@ -474,27 +499,84 @@ def imshow(image, vrange='indep1', zoom=1, title='', col_wrap=None, ax=None,
...
+        for i, img in enumerate(image_tmp):
+            if img.ndim == 2:
+                # make all of the RGB channels identical so it's grayscale
+                img = np.dstack([img, img, img])
+            if img.shape[-1] == 3:
+                # add an alpha channel of all 1s
+                img = np.dstack([img, np.ones_like(img[..., 0])])
+            image_tmp[i] = img
+        image_tmp = np.array(image_tmp)
+    return image_tmp, title_tmp, contains_rgb
```
Comment (billbrod) [inline pyrtools/tools/display.py:520]: This bit here is to support mixing grayscale, RGB, and RGBA images. I couldn't find a way to create an array containing an array of size, e.g., `(10, 10)` and `(10, 10, 4)` (though `(5,5)` and `(10,10,4)` is okay), so I expand the grayscale image into an RGBA image in order to make this work. I just duplicate the grayscale values across all three channels -- that is correct, right? That's what `skimage.color.gray2rgb` does (I don't use it because we don't have `skimage` as a requirement).

File: pyrtools/tools/display.py (hunk 3 above)
Comment (billbrod) [reply pyrtools/tools/display.py]: Alright, made those changes

File: pyrtools/tools/display.py (hunk 2 above)
Comment (billbrod) [reply pyrtools/tools/display.py]: Okay, so I just added this, but I'd like to get input from more people: this means we no longer support tensors at all. That's not bad, because, as I said, we're writing `plenoptic.imshow` to handle this, and tensors didn't always behave like you wanted them to. But it is a major change.

**Comment by billbrod** [comment]
- [ ] handle numpy `VisibleDeprecationWarning` about creating array from ragged nested list: if there are arrays of different sizes, set `dtype=np.object`
